### PR TITLE
Remove 'bio_data' from Tripal

### PR DESCRIPTION
**Pull request by spficklin**
# Tripal 4 Core Dev Task

Issue #1565 

## Description

The goal of this PR is to remove the use of the word `bio_data` in Tripal. As part of Tripal3 it was used as an equilvalent of `node` but for Tripal content types.   The goal was to use the bundle name (e.g. organism, analysis, contact, etc.) instread of `bio_data`.   After digging through the code I found this wasn't easily doable.  I got part of the way there, but hit on problems when creating brand new content items.  So, instead, I opted for using `tripal` instead of `bio_data` and I enabled the automatic creation of URL aliases using the `url_format` setting for content types.  So, all tripal entities will get a default URL of something like `tripal/1`, `tripal/2`, etc. but with the URL alias will appear to the user as something like `organism/1`, `project/2`, etc.  

The best benefit for this change is that content type are no longer referred to as `bio_data_1` and `bio_data_2` but rather by a more human-readable name (e.g. `organism`, `contact`, etc.).

## Testing?

There are two items to test.
1. Entity Creation: To  test this, simply update your Drupal installation to include this branch, install Drupal and create tripal an organism, project, analysis... whatever you like.  After creation, they should have URL aliases appropraite for the content type and you should be able to get to the same entity by changing the URL to have `tripal\<num>` where `<num>` is the same ID in the original URL (e..g `organism\1`)
2. Content Type Creation:  Now that we no longer use `bio_data` for the name of content types, you can create a new content type and it will use as a name a lowercase version of the label you give it.  Go to Admin > Structure > Tripal Content Types  and click the `Add Tripal Content Type` button at the top.  You should be able to create a new content type. Notice the machine name in the form (to the right of the label field), it should be a lowercase version of whatever you type in.  Any non alphanumeric values get converted to underscores.

### CODE REVIEW

**Comment by dsenalik** [comment]
@spficklin this error in your automated tests
`
2) Drupal\Tests\tripal\Functional\TripalRoutePermissionsTest::testTripalContentPages
Exception: The name is required when creating a TripalEntityType.
`
is a validation I just added in pull #1583 - when I was adding the name I thought it was missing here too https://github.com/tripal/tripal/blob/e7a20967bb21646bb9b6cc7da89e8cac77675821/tripal/src/Entity/TripalEntityType.php#L243-L245

If by chance it is valid to not have a name for an entity, then you can remove these three lines on your branch and my apologies 😦. If, on the other hand it is your test that is in error, you can fix that and I will be happy that it was caught 😀.

**Comment by laceysanderson** [comment]
I marked this as a draft PR until the tests are passing.

**Comment by spficklin** [comment]
@dsenalik the validation check you added is correct.  The problem with the testing is something else.  Digging into it also made me realize that we don't need a `name` parameter for the content types. Now that we no longer need to keep a numeric ID around (i.e. for the bio_data_x machine name) we no longer need a separte Id and Name.

**Comment by dsenalik** [comment]
@spficklin I was curious about why `bio_data` needed to be changed. It is not a big deal, this may be the only example I have, but I used a bio_data link to access some content and if it changes later when we update Tripal it would break the link.
bio_data-link

**Comment by spficklin** [comment]
Hey @dsenalik .  In Tripal v3 `bio_data_x` was used all over the place, in field names, as content types and it was really annoying to have to use a lookup table when you were digging around in the code or tables in postgres.    It made the code and data hard to follow.  Also, there was no consistency between the bio_data numbers and the content types after a site had prepared the initial content types. New types would not have consistent types across sites.   The changes in Tripal v4 allow us to use a human-readable name for the content types which is so much easier for field naming and looking through all those tables that Drupal creates for fields.    So, removing it is more convenience for us rather than anything related to functionality..

**Comment by laceysanderson** [comment]
Yes, there have been bugs in extension modules pop-up where they made an assumption on a specific bio_data_# always being a given content type until one site it wasn't 🙈 Also as Stephen mentioned, it just makes coding extension modules more difficult than it needed to be when trying to target specific content types.

For your link, the first iteration had a synonym key which was intended to provide alias' back to the old bio_data_# to keep it intact. Since the synonym was in the YAML, you could override it with the specific bio_data_# for your site and ensure links like the one you showed do not break 💞

**Comment by dsenalik** [comment]
Maybe there is some confusion, I was not referring to the `bio_data_#` but rather to the generic `/bio_data/entity#` in a URL. Such as `/Project/1234` is equivalent to `/bio_data/1234` in a URL, and it looks like this is what is changing to `/tripal/1234` (though there are some differences to this rule when editing, and in the web services)

The alias option is very desirable, because what if you put the link in a paper that now can't be changed.

I fully agree that looking at `bio_data_#` tables in the DB is always confusing 😕

**Comment by spficklin** [comment]
@dsenalik oh... for the URLs you're right, that's a change.   I just throught I'd get rid of it all together and sicne with Tripal v3 it should devault to using the URLs aliases I made the assumption that changing this wouldn't be a bad idea for published URLs.  But you're right, if someone published URLs with the bio_data in them they won't work anymore.  We could, as part of the site upgrade process, add as a secondary URL alias the old bio_data URLs from the Tripal v3 sites....

**Comment by spficklin** [comment]
Aside from the naming of the branch which I guess I munged, this PR is ready for review!  All tests are passing.

**Review by dsenalik — COMMENTED** [review]
As far a code review, this seems good.

For testing, I built a new docker on this branch with 
`docker build --tag=tripaldocker:testing-1565 --build-arg drupalversion="10.1.x-dev" --file tripaldocker/Dockerfile-php8.2-pgsql13 ./`

I created three content types and the URL created was as expected
```
/organism/1
/gene/2
/genetic_marker/3
```

I verified that the tripal URLs worked
```
/tripal/1
/tripal/2
/tripal/3
```

I verified that the bio_data URLs did not work
```
/bio_data/1
/bio_data/2
/bio_data/3
```

I created a new content type "Primer" under the "Genetic" category using the CV term "DNA primer (SO:0000112)"
**The machine name was not populated**
And if I save
`The website encountered an unexpected error. Please try again later.
Drupal\Core\Entity\EntityMalformedException: The entity does not have an ID. in Drupal\Core\Config\Entity\ConfigEntityStorage->save() (line 246 of core/lib/Drupal/Core/Config/Entity/ConfigEntityStorage.php).`

**Comment by spficklin** [comment]
Whoops!  Sorry about that @dsenalik   I borked  the form when I made fixes and forgot to check it.  You should be able to create those new entity types now.  I did notice you can't add fields to new entity types, but I don't think that's a problem with this PR, but should e added as a new issue. I'll do that.    It's ready for testing that last bit again.

**Comment by dsenalik** [comment]
I tested your changes, and it is now working as expected. I entered "Primer" as the content type, and the machine name was populated with "primer". I then clicked the edit link and changed it to "primer1" successfully.
primer1

**Comment by dsenalik** [comment]
I would like to discuss, maybe at the next cofest, changing the `bio_data` in the URL to `tripal` though. I feel this particular change could be a potential problem for existing sites or modules.

**Comment by spficklin** [comment]
@dsenalik Sure, we can wait to merge it if you want to chat on it.   I thnk though we can accommodate it, if we make sure to add a second URL alias, when upgrading a Tv3 site to Tv4 to set the `bio_data/x` URL alias.  That means both the new `/tripal/xxx` and previous `bio_data/yyyy` and the prefered `organism/xxx` will all still work.   We did something similar for the upgrade from Tv3 to Tv4.

**Comment by spficklin** [comment]
Oh, I should add. I didn't like setting it to `/tripal/xxxx` but I hated `/bio_data/yyyy`. I just wanted to get rid of it.  I suspect @laceysanderson might feel the same way?  Anway, if we're okay preserving `/bio_data/xxx` as a URL alias for upgraded sites then maybe we could come up with something better than `/tripal/xxxx`?   Drupal uses `/node/` for their content types.  Maybe we could come up with something better.

**Comment by dsenalik** [comment]
I agree that `tripal` is probably the most logical choice. And I can see you had a bad experience with `bio_data` 😆 Here's an idea, what about a single letter? `/t/1234` - although these are not URLs you are going to type in (usually) so it really makes no difference. This could potentially make debugging harder though, how can you grep for just `t`?
If we have and document the `bio_data` alias method for upgrading sites, then I am happy with this. I will mark my review as approved and let the will of the PMC prevail 😎

**Review by dsenalik — APPROVED** [review]
Working as intended 🎉

**Comment by spficklin** [comment]
I do like '/t/xxxx`.  It makes the URL shorter, it could also be `/data/xxxx`. I like that better than `/bio_data/xxxx`.  

> I will mark my review as approved and let the will of the PMC prevail 😎

Deferring to @laceysanderson then to decide if we wait for further discussion or merge.  @laceysanderson , if you vote to change the URL from `/tripal/xxxx` I can fix it here before we merge.

**Comment by laceysanderson** [comment]
To be honest I like any of `bio_data/xxxx` , `tripal/yyyy`, `data/zzzz` so I may not be the best tie-breaker 🤪 . Of these I might lean slightly towards `data/zzz` since it might be easier to search for if there is a problem as Tripal is literally everywhere in the code :zany_face:!

I think that we just let Stephen pick one rather than wait for discussion as this is a minor enough change to not need a lot of discussion. All I ask is that we create an issue to remind us to add the bio_data redirect in the upgrade code!

**Comment by dsenalik** [comment]
Okay @spficklin you get to decide! If I can, tomorrow I will create the issue for bio_data redirect, as I am out thursday-friday.

**Comment by dsenalik** [comment]
And don't forget `/t/xxx` or `/d/yyy` or `/🥕/zzz` as other options

**Comment by spficklin** [comment]
Okay. After thinking on it, I decided to just use `bio_data/xxxx`.  That way we don't have to create more work for ourselves by creating those aliases.  No reason to make more work for everyone if I'm the only one who really dislikes it.  Lacey makes a good point that it's easier to find where these links are used in the code if we don't use `tripal/xxxx`.  

I've updatd the code and running tests, and  i did another round of manual testing.  If all the automated tests pass I'll go ahead and merge.
